clippy_lints/src/undocumented_unsafe_blocks.rs

```diff
@@ -0,0 +1,142 @@
+            if !is_lint_allowed(cx, UNDOCUMENTED_UNSAFE_BLOCKS, block.hir_id);
+            if !in_external_macro(cx.tcx.sess, block.span);
+            if let BlockCheckMode::UnsafeBlock(UnsafeSource::UserProvided) = block.rules;
+            if let Some(enclosing_scope_hir_id) = cx.tcx.hir().get_enclosing_scope(block.hir_id);
+            then {
+                find_candidate(cx, block.span, enclosing_scope_hir_id)
+            }
+        }
+    }
+
+    fn check_local(&mut self, cx: &LateContext<'_>, local: &'_ Local<'_>) {
+        if_chain! {
+            if !is_lint_allowed(cx, UNDOCUMENTED_UNSAFE_BLOCKS, local.hir_id);
+            if !in_external_macro(cx.tcx.sess, local.span);
+            if let Some(init) = local.init;
+            if let ExprKind::Block(block, _) = init.kind;
+            if let BlockCheckMode::UnsafeBlock(UnsafeSource::UserProvided) = block.rules;
+            if let Some(enclosing_scope_hir_id) = cx.tcx.hir().get_enclosing_scope(block.hir_id);
+            then {
+                self.local = true;
+                find_candidate(cx, local.span, enclosing_scope_hir_id)
+            }
+        }
+    }
+}
+
+fn find_candidate(cx: &LateContext<'_>, span: Span, enclosing_hir_id: HirId) {
+    if let Some(true) = block_has_safety_comment(cx.tcx, enclosing_hir_id, span) {
+        return;
+    }
+
+    let block_indent = indent_of(cx, span);
+    let suggestion = format!("// Safety: ...\n{}", snippet(cx, span, ".."));
+
+    span_lint_and_sugg(
+        cx,
+        UNDOCUMENTED_UNSAFE_BLOCKS,
+        span,
+        "unsafe block missing a safety comment",
+        "consider adding a safety comment",
+        reindent_multiline(Cow::Borrowed(&suggestion), true, block_indent).to_string(),
+        Applicability::HasPlaceholders,
+    );
+}
+
+fn block_has_safety_comment(tcx: TyCtxt<'_>, enclosing_hir_id: HirId, block_span: Span) -> Option<bool> {
+    let map = tcx.hir();
+    let source_map = tcx.sess.source_map();
+
+    let enclosing_scope_span = map.opt_span(enclosing_hir_id)?;
+    let between_span = enclosing_scope_span.until(block_span);
+
+    let file_name = source_map.span_to_filename(between_span);
+    let source_file = source_map.get_source_file(&file_name)?;
+
+    let lex_start = (between_span.lo().0 + 1) as usize;
+    let mut src_str = source_file.src.as_ref()?[lex_start..between_span.hi().0 as usize].to_string();
+
+    src_str.retain(|c| !c.is_whitespace());
+
+    let src_str_split = src_str.lines().rev().collect::<Vec<_>>();
+    let src_str = src_str_split.join("");
+
+    let mut pos = 0;
+    let mut found_safety_comment = false;
+
+    for token in rustc_lexer::tokenize(&src_str) {
+        match token.kind {
+            TokenKind::LineComment { doc_style: None }
+            | TokenKind::BlockComment { doc_style: None, terminated: true } => {
+                if src_str[pos + 2.. pos + token.len].to_ascii_uppercase().contains("SAFETY:") {
+                    found_safety_comment = true;
```

Comment:
Please also check if the safety comment is just `...`, i.e. if the person just jammed in the suggestion without actually typing out any safety comment, this'd prevent laziness and false positives from shoddy lint wack-a-mole-ing.